### PR TITLE
add first TS post

### PR DESCRIPTION
**Pull request by skeydan**

### CODE REVIEW

**Review by batpigandme — COMMENTED** [review]
My aim here is simplification, since this is your “introduction to time series forecasting with `torch`.” My goal is to make the post maximally useful to someone who's using this to figure out their own work (make it more skimable), or even someone reading it through!

**Comment by batpigandme** [comment]
GitHub is collapsing comments, and they're showing up in kind of an odd order. So, it might be easiest to go through them from the `.Rmd` view.

File: _posts/2021-03-10-forecasting-time-series-with-torch_1/forecasting_1.Rmd [hunk 1]
```diff
@@ -0,0 +1,633 @@
+---
+title: "Introductory time series forecasting with torch"
+description: > 
+ This post is an introduction to time series forecasting with torch. Central topics are data input on the one hand and practical usage of RNNs (GRUs/LSTMs) on the other. Upcoming posts will build on this and introduce increasingly involved architectures.
```
[the remark this answers is not in the document]
Comment (skeydan) [reply _posts/2021-03-10-forecasting-time-series-with-torch_1/forecasting_1.Rmd:4]: yeah, I see what you mean :-) It was a conflict of rhythm vs. content ;-) 
But then, rhythm didn't get worse really!

File: _posts/2021-03-10-forecasting-time-series-with-torch_1/forecasting_1.Rmd [hunk 2]
```diff
@@ -0,0 +1,633 @@
...
+    self_contained: false
+    toc: true
+preview: images/d.jpg
+---
+
+```{r setup, include=FALSE}
+knitr::opts_chunk$set(echo = TRUE, eval = FALSE)
+```
+
+This is the first post in a series introducing time series forecasting with `torch`. It does assume some prior experience with `torch` and/or deep learning. But as far as time series are concerned, it starts right from the beginning: using recurrent neural networks (GRU or LSTM) to predict how something develops in time.
```
[the remark this answers is not in the document]
Comment (skeydan) [reply _posts/2021-03-10-forecasting-time-series-with-torch_1/forecasting_1.Rmd:27]: ok... the colon was intended to more clearly convey that this really is the most basic way to do it... but I guess it does stay clear even with the comma...

File: _posts/2021-03-10-forecasting-time-series-with-torch_1/forecasting_1.Rmd [hunk 3]
```diff
@@ -0,0 +1,633 @@
...
+
+In this post, we build a network that uses a sequence of observations to predict a value for the very next point in time. What if we'd like to forecast a *sequence* of values, corresponding to, say, a week or a month of measurements?
+
+One thing we could do is feed back into the system the previously forecasted value; this is something we'll try at the end of this post. Subsequent posts will explore other options, some of them involving significantly more complex architectures. It will be interesting to compare their performances; but the essential goal is to introduce some `torch` "recipes" that you can, with appropriate modifications, apply to your own data.
```
[the remark this answers is not in the document]
Comment (skeydan) [reply _posts/2021-03-10-forecasting-time-series-with-torch_1/forecasting_1.Rmd:31]: > this is implied by recipes
I'm always amazed by this type of judgment (?) .... Not flattering when you see it but so completely true

File: _posts/2021-03-10-forecasting-time-series-with-torch_1/forecasting_1.Rmd [hunk 4]
```diff
@@ -0,0 +1,633 @@
...
+
+    Rows: 52,608
+    Columns: 5
+    $ Time        <dttm> 2012-01-01 00:00:00, 2012-01-01 00:30:00, 2012-01-01 01:00:00,…
+    $ Demand      <dbl> 4382.825, 4263.366, 4048.966, 3877.563, 4036.230, 3865.597, 369…
+    $ Temperature <dbl> 21.40, 21.05, 20.70, 20.55, 20.40, 20.25, 20.10, 19.60, 19.10, …
+    $ Date        <date> 2012-01-01, 2012-01-01, 2012-01-01, 2012-01-01, 2012-01-01, 20…
+    $ Holiday     <lgl> TRUE, TRUE, TRUE, TRUE, TRUE, TRUE, TRUE, TRUE, TRUE, TRUE, TRU…
+
+Depending on what subset of variables is used, and whether and how data is temporally aggregated, these data may serve to illustrate a variety of different techniques. For example, in the third edition of [Forecasting: Principles and Practice](https://otexts.com/fpp3/) daily averages are used to teach [quadratic regression with ARMA errors](https://otexts.com/fpp3/forecasting.html). In this first introductory post though, as well as in most of its successors, we'll attempt to forecast `Demand` without relying on additional information, and we keep the original resolution.
```
[the remark this answers is not in the document]
Comment (skeydan) [reply _posts/2021-03-10-forecasting-time-series-with-torch_1/forecasting_1.Rmd:58]: Yes, I see what you mean. But it gets taken up later, when I say how difficult it is to forecast the univariate series without relying on external information. (So, it is there for a reason - to set up an excuse early, if you want ;-))

File: _posts/2021-03-10-forecasting-time-series-with-torch_1/forecasting_1.Rmd [hunk 5]
```diff
@@ -0,0 +1,633 @@
...
+
+To get an impression of how electricity demand varies over different timescales, let's inspect data for two months that nicely illustrate the U-shaped relationship between temperature and demand: January, 2014 and July, 2014. (It's true that we won't make use of temperature in most models, but this is worthwhile looking at in any case.)
```
[the remark this answers is not in the document]
Comment (skeydan) [reply _posts/2021-03-10-forecasting-time-series-with-torch_1/forecasting_1.Rmd:60]: ok! (hard to do but right, of course)

File: _posts/2021-03-10-forecasting-time-series-with-torch_1/forecasting_1.Rmd [hunk 6]
```diff
@@ -0,0 +1,633 @@
...
+  theme_minimal()
+
+```
+
+```{r, eval = TRUE, echo = FALSE, layout="l-page", fig.cap="Temperature and electricity demand (normalized). Victoria, Australia, 07/2014."}
+knitr::include_graphics("images/july_2014.png")
+
+```
+
+It's winter; temperature fluctuates below average, while electricity demand is above average (heating). There is strong variation over the day; we see troughs in the demand curve corresponding to ridges in the temperature graph, and vice versa. While diurnal variation dominates, there is hebdomadal variation as well. Between weeks though, we don't see much difference.
```
[the remark this answers is not in the document]
Comment (skeydan) [reply _posts/2021-03-10-forecasting-time-series-with-torch_1/forecasting_1.Rmd:84]: > which is oddly verbose in English)
no, i like it, in fact (rhythm, again ;-))

File: _posts/2021-03-10-forecasting-time-series-with-torch_1/forecasting_1.Rmd [hunk 7]
```diff
@@ -0,0 +1,633 @@
...
+
+-   In addition to the highly important trend component, the data are characterized by multiple levels of seasonality.
+
+Of these, the latter is less of a problem for the techniques we're working with here. If we found that some level of seasonality went undetected, we could try to adapt the current configuration in a number of uncomplicated ways:
+
+-   Use an LSTM instead of a GRU. In theory, LSTM should better be able to capture additional lower-frequency components due to its secondary storage, the cell state.
+
+-   Stack multiple layers of GRU/LSTM. In theory, this should allow for learning a hierarchy of temporal features, analogously to what we see in a convolutional neural network.
+
+To address the former obstacle, bigger changes to the architecture would be needed. We may attempt to do that in a later, "bonus", post. But in the upcoming installments, we'll first dive into often-used techniques for sequence prediction, also porting to numerical time series things that are commonly done in natural language processing.
```
[the remark this answers is not in the document]
Comment (skeydan) [reply _posts/2021-03-10-forecasting-time-series-with-torch_1/forecasting_1.Rmd:631]: true! Perhaps I can do that in a footnote, then, later, to not disrupt the original flow

File: _posts/2021-03-10-forecasting-time-series-with-torch_1/forecasting_1.Rmd [hunk 8]
```diff
@@ -0,0 +1,633 @@
...
+knitr::include_graphics("images/july_stl.png")
+```
+
+And here, for January:
+
+```{r, eval = TRUE, echo = FALSE, layout="l-page", fig.cap="STL decomposition of electricity demand. Victoria, Australia, 01/2014."}
+knitr::include_graphics("images/jan_stl.png")
+```
+
+Both nicely illustrate the strong circadian and weekly seasonalities (with diurnal variation substantially stronger in January). If we look closely, we can even see how the trend component is more influential in January than in July. This again hints at much stronger difficulties predicting the January than the July developments.
```
[the remark this answers is not in the document]
Comment (skeydan) [reply _posts/2021-03-10-forecasting-time-series-with-torch_1/forecasting_1.Rmd:126]: again, gathering excuses upfront ;-)))

File: _posts/2021-03-10-forecasting-time-series-with-torch_1/forecasting_1.Rmd [hunk 9]
```diff
@@ -0,0 +1,633 @@
...
+     0.9399
+    ... [the output was truncated (use n=-1 to disable)]
+    [ CPUFloatType{336,1} ]
+
+    $y
+    torch_tensor
+    -0.6771
+    [ CPUFloatType{1} ]
+
+Yes: This is what we wanted to see. The input sequence has `n_timesteps` values in the first dimension, and a single one in the second, corresponding to the only feature present, `Demand`. And also as intended, the prediction tensor holds a single value only, corresponding to -- as we know -- `n_timesteps+1`.
```
[the remark this answers is not in the document]
Comment (skeydan) [reply _posts/2021-03-10-forecasting-time-series-with-torch_1/forecasting_1.Rmd:252]: It does!

File: _posts/2021-03-10-forecasting-time-series-with-torch_1/forecasting_1.Rmd [hunk 10]
```diff
@@ -0,0 +1,633 @@
...
+valid_ds <- elec_dataset(elec_valid, n_timesteps)
+valid_dl <- valid_ds %>% dataloader(batch_size = batch_size)
+
+test_ds <- elec_dataset(elec_test, n_timesteps)
+test_dl <- test_ds %>% dataloader(batch_size = 1)
+```
+
+# Model
+
+The model consists of an RNN -- of type GRU or LSTM, as per the user's choice -- and an output layer. The RNN does most of the work; the single-neuron linear layer that outputs the prediction just compresses its vector input to a single value.
```
[the remark this answers is not in the document]
Comment (skeydan) [reply _posts/2021-03-10-forecasting-time-series-with-torch_1/forecasting_1.Rmd:300]: I should grep for "just" before pushing ... will do next time ;-))

File: _posts/2021-03-10-forecasting-time-series-with-torch_1/forecasting_1.Rmd [hunk 11]
```diff
@@ -0,0 +1,633 @@
...
+
+Here first is the model definition; read on for a more detailed explanation.
```
[the remark this answers is not in the document]
Comment (skeydan) [reply _posts/2021-03-10-forecasting-time-series-with-torch_1/forecasting_1.Rmd:302]: > my lack of technical savvy shows
Haha, now I got you ;-))) (the equivalent of my "just"s, I mean ;-))

I think my intention was to assure readers that what comes next will be explained. ("don't be scared")

But then again, you're right, that's not too unusual in a blog post, that stuff gets explained.

File: _posts/2021-03-10-forecasting-time-series-with-torch_1/forecasting_1.Rmd [hunk 12]
```diff
@@ -0,0 +1,633 @@
...
+preview: images/d.jpg
+---
+
+```{r setup, include=FALSE}
+knitr::opts_chunk$set(echo = TRUE, eval = FALSE)
+```
+
+This is the first post in a series introducing time series forecasting with `torch`. It does assume some prior experience with `torch` and/or deep learning. But as far as time series are concerned, it starts right from the beginning: using recurrent neural networks (GRU or LSTM) to predict how something develops in time.
+
+In this post, we build a network that uses a sequence of observations to predict a value for the very next point in time. What if we'd like to forecast a *sequence* of values, corresponding to, say, a week or a month of measurements?
```
[the remark this answers is not in the document]
Comment (skeydan) [reply _posts/2021-03-10-forecasting-time-series-with-torch_1/forecasting_1.Rmd:29]: I think what I'm trying to do here is answer an anticipated question: "Who wants to predict just 30 minutes ahead?"
For the linking of posts once they're up, yes, that sounds good, will do. But I'm not sure I want to formally announce them here already... (and unless people already are familiar with those techniques, they will remain purely abstract anyway...)

File: _posts/2021-03-10-forecasting-time-series-with-torch_1/forecasting_1.Rmd [hunk 13]
```diff
@@ -0,0 +1,633 @@
...
+knitr::include_graphics("images/jan_stl.png")
+```
+
+Both nicely illustrate the strong circadian and weekly seasonalities (with diurnal variation substantially stronger in January). If we look closely, we can even see how the trend component is more influential in January than in July. This again hints at much stronger difficulties predicting the January than the July developments.
+
+Now that we have an idea what awaits us, let's begin by creating a `torch` `dataset`.
+
+# Data input
+
+For that purpose, we need to be able to clearly formulate what we intend to do.
```
[the remark this answers is not in the document]
Comment (skeydan) [reply _posts/2021-03-10-forecasting-time-series-with-torch_1/forecasting_1.Rmd:132]: right. I know what I wanted to convey, but the way I'm doing it is clumsy. Will change.

File: _posts/2021-03-10-forecasting-time-series-with-torch_1/forecasting_1.Rmd [hunk 14]
```diff
@@ -0,0 +1,633 @@
...
+
+[^1]: By default means that if the optional `return_sequences` argument is not passed. See [here](https://blogs.rstudio.com/ai/posts/2020-12-17-torch-convlstm/#interlude-outputs-states-hidden-values-whats-what) for a systematic comparison of RNN return values for both `torch` and Keras.
+
+In later posts, we will make use of more than the last time step. Sometimes, we'll use the sequence of hidden states (the $h$s) instead of the outputs (the $o$s). So you may feel like asking, what if we used $h_{t+1}$ here instead of $o_{t+1}$? The answer is: With a GRU, this would not make a difference, as those two are identical. With LSTM though, it would, as LSTM keeps a second, namely, the "cell", state[^2].
+
+[^2]: Again, see [this post](https://blogs.rstudio.com/ai/posts/2020-12-17-torch-convlstm/#interlude-outputs-states-hidden-values-whats-what) for more details.
+
+Now that we've talked about `forward()`, less is to be said about `initialize()`. For ease of experimentation, we instantiate either a GRU or an LSTM based on user input. Two things are worth noting:
```
[the remark this answers is not in the document]
Comment (skeydan) [reply _posts/2021-03-10-forecasting-time-series-with-torch_1/forecasting_1.Rmd:393]: ok

**Comment by skeydan** [comment]
thanks a thousand times Mara, that was super helpful!!
(((And I love how things get not just more clear, that anyway, but also, more terse as well ;-)))

**Pull request merged**